### ISSUE TITLE
Action Bar: detect HTML for word count

### ISSUE THREAD

**Issue by mobitar** [standardnotes/bounties#40]
When using the Plus or Bold editor, Action Bar does not understand that the text is HTML. So the word count is invalid. We'll want to detect HTML, and then strip the markup and do a word count on the actual contents.

Regex might be one option, but might impact performance. We might just be able to do a simple prefix check of "<html>" or something similar.

https://github.com/sn-extensions/action-bar

**Comment by bryvin** [comment]
Sounds like using Regex to parse HTML is not a good idea, though it may work "good enough"

https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454

**Comment by mobitar** [comment]
We have actually been using Regex with FileSafe with no issues: https://github.com/standardnotes/editor-kit/blob/master/src/lib/EditorKitInternal.js#L150

**Comment by bryvin** [comment]
Gotcha, it seems like it can be used to test if a document contains HTML without failure; however to pull the text out of the tags may be a different story. 

Though using Regex is still likely possible to use for a much better solution than the current counting of HTML but it may not be 100% accurate in all cases. Definitely worth a shot though!

**Comment by mobitar** [comment]
I think we also pull out the plain text from HTML when we go to generate previews, so something similar can be used: https://github.com/standardnotes/editor-kit/blob/master/src/lib/Util.js#L2

**Comment by bryvin** [comment]
I'll give this a shot today then; will update later

**Comment by bryvin** [comment]
I tried using the same method of getting the textContent that the previews use as suggested; unfortunately the streamed content from SN doesn't include traditional line breaks. It's only when the editor (tested with Bold Editor) inits that it formats the content of the note to include the line breaks. But as soon as the user edits the note within the editor it breaks the count until a re-init happens.

I have recorded this behavior:
![test](https://user-images.githubusercontent.com/5969300/64552802-a3500500-d2fd-11e9-8ad9-9c3770896693.gif)

**Comment by mobitar** [comment]
Not sure I'm following. When the editor inits and formats the note, wouldn't the action bar be notified of this immediately?

**Comment by bryvin** [comment]
The video I recorded is pretty trash sorry.

What happens is when the editor is initialized it does it's own parsing and formatting of the note text, which is what goes into the displayed version. This version is what you'd expect a traditional html file to look like with line breaks after tags; essentially not minified.

So when that happens, it parses correctly with textContent as the note text actually contains those line breaks as well (after the editor has parsed it on init and replaces the note's content hard to tell but in my video after the page goes white in the top right you can quickly see "Saving..." which is the editor saving the formatted html).

The problem with all of this though is that the editor doesn't maintain the formatted html after that initial init. So when you edit the note's content/text within the editor (as shown in my video) it doesn't add line breaks or format the html; it just has basically a minified version of the html.

Since textContent relies on the _actual_ line breaks to be there to return the correct string this is broken. Thus the counting of paragraphs is also broken because there are no line breaks at all in the returned string.

Basically it needs some more complex reading of the document if it's an html version. An actual parser I guess.

**Comment by mobitar** [comment]
I'd say paragraph count is less important than word count. Could we detect that we can't properly detect paragraph count, and show "N/A" for the count instead of a number?

**Comment by bryvin** [comment]
For some reason the example I had also never updated the word count more than 1, even with the initial loading; but I'll mess around with this some more and see what I can come up with.

**Comment by bryvin** [comment]
Did some more work on this and have gotten it to a point where in the Bold Editor it was behaving _almost_ as one would expect.. but not quite. Then switched to the Plus Editor and completely different results.

So this appears to actually be very hard, for all the stats: words, characters and paragraphs. 

I then went back to the Bold Editor and found out it works when you paste stuff in (exact match to what some online resources show for counts), but then when you start typing it changes the formatting provided by textContent again and it has quirky behavior... so this actually seems to be more difficult and may require some 3rd party resource to parse the HTML and provide a standardized format to use for analyzing. I know you stance on that so didn't want to go looking for options there, but let me know your thoughts.
